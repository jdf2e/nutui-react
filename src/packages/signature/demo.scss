.demo-tips {
  color: #1a1a1a;
}
