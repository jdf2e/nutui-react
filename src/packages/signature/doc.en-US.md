# Signature 

### Intro

Signature component based on canvas.

### Install

```javascript
// react
import { Signature } from '@nutui/nutui-react'

```

## Demo

### Basic usage

:::demo

```tsx
import  React from "react";
import { Signature} from '@nutui/nutui-react';

const App = () => {
    const confirm = (canvas: HTMLCanvasElement, data: string) => {
      const img = document.createElement('img')
      img.src = data
      const demo = document.querySelector('.demo') as HTMLElement
      demo.appendChild(img)
    }
    const clear = () => {
      const img = document.querySelector('.demo img')
      if (img) {
        img.remove()
      }
    }
    const demoStyles: React.CSSProperties = { margin: '1em 0' }
    return (
    <>
        <Signature onConfirm={confirm} onClear={clear} />
        <p className="demo-tips demo" style={demoStyles}>
           Tips: click the confirm button, and the signature image is displayed below
        </p>
    </>
    );
};
export default App;
```

:::

### Modify color and signature thickness

:::demo

```tsx
import  React from "react";
import { Signature} from '@nutui/nutui-react';

const App = () => {
    const confirm = (canvas: HTMLCanvasElement, data: string) => {
    const img = document.createElement('img')
    img.src = data
    const demo = document.querySelector('.demo') as HTMLElement
    demo.appendChild(img)
    }
    const clear = () => {
    const img = document.querySelector('.demo img')
    if (img) {
        img.remove()
    }
    }
    const demoStyles: React.CSSProperties = { margin: '1em 0' }
    return (
    <>
        <Signature
            lineWidth={4}
            strokeStyle="green"
            onConfirm={confirm}
            onClear={clear}
         />
        <p className="demo-tips demo" style={demoStyles}>
            Tips: click the confirm button, and the signature image is displayed below
        </p>
    </>
    );
};
export default App;
```

:::

## API

### Props

| Attribute           | Description                           | Type   | Default                                              |
| -------------- | ------------------------------ | ------ | --------------------------------------------------- |
| customClass   |  Custom class                 | String | -                                                   |
| lineWidth     | Width of line                    | Number | 3                                                   |
| strokeStyle   | Drawing stroke color                   | String | '#000'                                              |
| type           | Picture format                       | String | 'png'                                               |
| unSupportTpl |  Display copy without canvas | String | 'sorry, the current browser doesn't support canvas, so we can't use this control! ' |

## Event

| 字段    | 说明                         | 回调参数                         |
| ------- | ---------------------------- | -------------------------------- |
| confirm`v1.3.8 废弃` | Click the confirm button to trigger the event callback function | Canvas and data URI displayed by signature imageURI |
| clear`v1.3.8 废弃`   | Click the re sign button to trigger the event callback function | -                               |
| onConfirm`v1.3.8` | Click the confirm button to trigger the event callback function | Canvas and data URI displayed by signature imageURI |
| onClear`v1.3.8`   | Click the re sign button to trigger the event callback function | -                               |

