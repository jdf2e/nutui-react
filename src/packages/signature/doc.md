# Signature 签名

### 介绍

基于 Canvas 的签名组件

### 安装

```javascript
import { Signature } from '@nutui/nutui-react'
```

## 代码演示

### 基础用法

:::demo

```tsx
import  React from "react";
import { Signature} from '@nutui/nutui-react';

const App = () => {
    const confirm = (canvas: HTMLCanvasElement, data: string) => {
      let img = document.createElement('img')
      img.src = data
      const demo = document.querySelector('.demo') as HTMLElement
      demo.appendChild(img)
    }
    const clear = () => {
      let img = document.querySelector('.demo img')
      if (img) {
        img.remove()
      }
    }
    const demoStyles: React.CSSProperties = { margin: '1em 0' }
    return (
    <>
        <Signature confirm={confirm} clear={clear}></Signature>
        <p className="demo-tips demo" style={demoStyles}>
            Tips: 点击确认按钮,下方显示签名图片
        </p>
    </>
    );
};
export default App;
```

:::

### 修改颜色和签字粗细

:::demo

```tsx
import  React from "react";
import { Signature} from '@nutui/nutui-react';

const App = () => {
    const confirm = (canvas: HTMLCanvasElement, data: string) => {
    let img = document.createElement('img')
    img.src = data
    const demo = document.querySelector('.demo') as HTMLElement
    demo.appendChild(img)
    }
    const clear = () => {
    let img = document.querySelector('.demo img')
    if (img) {
        img.remove()
    }
    }
    const demoStyles: React.CSSProperties = { margin: '1em 0' }
    return (
    <>
        <Signature
            lineWidth={4}
            strokeStyle={'green'}
            confirm={confirm}
            clear={clear}
        ></Signature>
        <p className="demo-tips demo" style={demoStyles}>
            Tips: 点击确认按钮,下方显示签名图片
        </p>
    </>
    );
};
export default App;
```

:::

## API

### Props

| 参数           | 说明                           | 类型   | 默认值                                              |
| -------------- | ------------------------------ | ------ | --------------------------------------------------- |
| custom-class   | 自定义 class                   | String | -                                                   |
| line-width     | 线条的宽度                     | Number | 3                                                   |
| stroke-style   | 绘图笔触颜色                   | String | '#000'                                              |
| type           | 图片格式                       | String | 'png'                                               |
| un-support-tpl | 不支持 Canvas 情况下的展示文案 | String | '对不起，当前浏览器不支持 Canvas，无法使用本控件！' |

## Event

| 字段    | 说明                         | 回调参数                         |
| ------- | ---------------------------- | -------------------------------- |
| confirm | 点击确认按钮触发事件回调函数 | canvas 和签名图片展示的 data URI |
| clear   | 点击重签按钮触发事件回调函数 | 无                               |
