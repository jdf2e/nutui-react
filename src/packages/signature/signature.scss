.nut-signature {
  .spcanvas_WEAPP {
    width: 100%;
    height: 100%;

    Canvas {
      width: 100%;
    }
  }

  &__inner {
    background-color: #fff;

    Canvas {
      background-color: #fff;
    }

    height: $signature-height;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  &__unsupport {
    font-size: $font-size-base;
  }
}
