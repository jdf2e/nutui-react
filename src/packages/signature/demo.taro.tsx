import './demo.scss'
import React from 'react'
import Taro from '@tarojs/taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import Header from '@/sites/components/header'
import Demo1 from './demos/taro/demo1'
import Demo2 from './demos/taro/demo2'

const SignatureDemo = () => {
  const [translated] = useTranslate({
    'zh-CN': {
      basic: '基础用法',
      title: '修改颜色和签字粗细',
    },
    'zh-TW': {
      basic: '基础用法',
      title: '修改顏色和簽字粗細',
    },
    'en-US': {
      basic: 'Basic Usage',
      title: 'Modify color and signature thickness',
    },
  })

  return (
    <>
      <Header />
      <div
        className={`demo demo-signature demo-taro-signature ${
          Taro.getEnv() === 'WEB' ? 'web' : ''
        }`}
      >
        <h2>{translated.basic}</h2>
        <Demo1 />
        <h2> {translated.title}</h2>
        <Demo2 />
      </div>
    </>
  )
}

export default SignatureDemo
