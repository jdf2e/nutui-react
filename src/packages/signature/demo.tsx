import React from 'react'
import './demo.scss'
import { Signature } from './signature'
import { useTranslate } from '../../sites/assets/locale'

interface T {
  basic: string
  title: string
  tips: string
}

const SignatureDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      basic: '基本用法',
      title: '修改颜色和签字粗细',
      tips: 'Tips: 点击确认按钮,下方显示签名图片',
    },
    'zh-TW': {
      basic: '基本用法',
      title: '修改顏色和簽字粗細',
      tips: 'Tips: 點擊確認按鈕,下方顯示簽名圖片',
    },
    'en-US': {
      basic: 'Basic Usage',
      title: 'Modify color and signature thickness',
      tips: 'Tips: click the confirm button, and the signature image is displayed below',
    },
  })
  const confirm = (canvas: HTMLCanvasElement, data: string) => {
    const img = document.createElement('img')
    img.src = data
    const demo = document.querySelector('.demo1') as HTMLElement
    demo.appendChild(img)
  }
  const clear = () => {
    const img = document.querySelector('.demo1 img')
    if (img) {
      img.remove()
    }
  }

  const confirm1 = (canvas: HTMLCanvasElement, data: string) => {
    const img = document.createElement('img')
    img.src = data
    const demo = document.querySelector('.demo2') as HTMLElement
    demo.appendChild(img)
  }

  const clear1 = () => {
    const img = document.querySelector('.demo2 img')
    if (img) {
      img.remove()
    }
  }

  const demoStyles: React.CSSProperties = { margin: '1em 0' }
  return (
    <>
      <div className="demo">
        <h2>{translated.basic}</h2>
        <Signature onConfirm={confirm} onClear={clear} />
        <p className="demo-tips demo1" style={demoStyles}>
          {translated.tips}
        </p>
        <h2> {translated.title}</h2>
        <Signature
          lineWidth={4}
          strokeStyle="green"
          onConfirm={confirm1}
          onClear={clear1}
        />
        <p className="demo-tips demo2" style={demoStyles}>
          {translated.tips}
        </p>
      </div>
    </>
  )
}

export default SignatureDemo
