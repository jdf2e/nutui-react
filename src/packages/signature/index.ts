import { Signature } from './signature'

export type { SignatureProps, FileType } from './signature'
export default Signature
