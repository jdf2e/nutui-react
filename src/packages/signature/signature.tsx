import React, {
  useRef,
  useState,
  useEffect,
  ForwardRefRenderFunction,
  useImperativeHandle,
  ReactNode,
} from 'react'
import { useConfig } from '@/packages/configprovider'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'
import { canUseDom } from '@/utils/can-use-dom'

export type SignatureType = 'jpg' | 'png'

export interface SignatureProps extends BasicComponent {
  type: SignatureType
  lineWidth: number
  strokeStyle: string
  unsupported: ReactNode
  onConfirm?: (
    canvas: HTMLCanvasElement,
    dataurl: string,
    hasSigatured?: boolean
  ) => void
  onClear?: () => void
}

const defaultProps = {
  ...ComponentDefaults,
  type: 'png',
  lineWidth: 2,
  strokeStyle: '#1a1a1a',
  unsupported: '',
} as SignatureProps
const InternalSignature: ForwardRefRenderFunction<
  unknown,
  Partial<SignatureProps>
> = (props, ref) => {
  const { locale } = useConfig()
  const {
    type,
    lineWidth,
    strokeStyle,
    unsupported,
    className,
    style,
    onConfirm,
    onClear,
    ...rest
  } = {
    ...defaultProps,
    ...props,
  }
  const classPrefix = `nut-signature`
  const canvasRef = useRef<HTMLCanvasElement>(null)
  const wrapRef = useRef<HTMLDivElement>(null)
  const [canvasHeight, setCanvasHeight] = useState(0)
  const [canvasWidth, setCanvasWidth] = useState(0)
  const ctx = useRef<CanvasRenderingContext2D | null>(null)
  const checkCanvas = () => {
    const elem = document.createElement('canvas')
    return !!(elem.getContext && elem.getContext('2d'))
  }
  const [isCanvasSupported, setIsCanvasSupported] = useState(false)
  const [hasSigatured, setSigatured] = useState(false)

  const isSupportTouch = canUseDom ? 'ontouchstart' in window : false
  const events = isSupportTouch
    ? ['touchstart', 'touchmove', 'touchend', 'touchleave']
    : ['mousedown', 'mousemove', 'mouseup', 'mouseleave']

  useEffect(() => {
    setIsCanvasSupported(checkCanvas)
  }, [])

  useEffect(() => {
    if (isCanvasSupported && canvasRef.current && wrapRef.current) {
      ctx.current = canvasRef.current.getContext('2d')
      setCanvasWidth(wrapRef.current.offsetWidth)
      setCanvasHeight(wrapRef.current.offsetHeight)
      addEvent()
    }
  }, [isCanvasSupported])

  const startEventHandler = (event: any) => {
    event.preventDefault()
    setSigatured(true)
    if (ctx.current && canvasRef.current) {
      ctx.current.beginPath()
      ctx.current.lineWidth = lineWidth as number
      ctx.current.strokeStyle = strokeStyle as string
      canvasRef.current.addEventListener(events[1], moveEventHandler, false)
      canvasRef.current.addEventListener(events[2], endEventHandler, false)
      canvasRef.current.addEventListener(events[3], leaveEventHandler, false)
    }
  }

  const addEvent = () => {
    if (canvasRef.current) {
      canvasRef.current.addEventListener(events[0], startEventHandler, false)
    }
  }

  const moveEventHandler = (event: any) => {
    event.preventDefault()

    const evt = isSupportTouch ? event.touches[0] : event
    if (canvasRef.current && ctx.current) {
      const coverPos = canvasRef.current.getBoundingClientRect()
      const mouseX = evt.clientX - coverPos.left
      const mouseY = evt.clientY - coverPos.top

      ctx.current.lineTo(mouseX, mouseY)
      ctx.current.stroke()
    }
  }

  const endEventHandler = (event: any) => {
    event.preventDefault()
    if (canvasRef.current) {
      canvasRef.current.removeEventListener(events[1], moveEventHandler, false)
      canvasRef.current.removeEventListener(events[2], endEventHandler, false)
    }
  }
  const leaveEventHandler = (event: any) => {
    event.preventDefault()
    if (canvasRef.current) {
      canvasRef.current.removeEventListener(events[1], moveEventHandler, false)
      canvasRef.current.removeEventListener(events[2], endEventHandler, false)
    }
  }
  const handleClearBtn = () => {
    setSigatured(false)
    if (canvasRef.current && ctx.current) {
      canvasRef.current.addEventListener(events[2], endEventHandler, false)
      ctx.current.clearRect(0, 0, canvasWidth, canvasHeight)
      ctx.current.closePath()
    }
    onClear && onClear()
  }

  const onSave = (canvas: HTMLCanvasElement) => {
    let dataurl = ''
    if (!hasSigatured) {
      onConfirm && onConfirm(canvas, dataurl as string, hasSigatured)
      return
    }
    switch (type) {
      case 'png':
        dataurl = canvas.toDataURL('image/png')
        break
      case 'jpg':
        dataurl = canvas.toDataURL('image/jpeg', 0.8)
        break
      default:
        dataurl = canvas.toDataURL('image/png')
    }
    onConfirm && onConfirm(canvas, dataurl as string, hasSigatured)
  }

  useImperativeHandle(ref, () => ({
    confirm: () => {
      onSave(canvasRef.current as HTMLCanvasElement)
    },
    clear: () => {
      handleClearBtn()
    },
  }))
  return (
    <div className={`${classPrefix} ${className}`} style={style} {...rest}>
      <div className={`${classPrefix}-inner`} ref={wrapRef}>
        <>
          {/* eslint-disable-next-line no-nested-ternary */}
          {isCanvasSupported ? (
            <canvas ref={canvasRef} height={canvasHeight} width={canvasWidth} />
          ) : unsupported ? (
            <>{unsupported}</>
          ) : (
            <p className={`${classPrefix}-unsupport`}>
              {locale.signature.unsupported}
            </p>
          )}
        </>
      </div>
    </div>
  )
}

export const Signature = React.forwardRef<unknown, Partial<SignatureProps>>(
  InternalSignature
)
Signature.displayName = 'NutSignature'
