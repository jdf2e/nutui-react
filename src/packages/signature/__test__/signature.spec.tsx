import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import 'vitest-canvas-mock'

import { Signature } from '../signature'
import Button from '@/packages/button'

test('props test', async () => {
  const App = () => {
    const confirm = (
      canvas: HTMLCanvasElement,
      dataurl: string,
      isSigned?: boolean
    ) => {
      console.log('confirm', isSigned, dataurl)
    }
    const clear = () => {
      console.log('clear')
    }
    const signatureRef = React.useRef<any>(null)

    return (
      <>
        <Signature
          className="test-signature"
          lineWidth={4}
          strokeStyle="green"
          ref={signatureRef}
          onConfirm={confirm}
          onClear={clear}
        />
        <div className="demo-btn">
          <Button
            type="default"
            size="small"
            style={{ margin: 8 }}
            data-testid="emit-click-1"
            onClick={() => signatureRef.current?.clear()}
          >
            重签
          </Button>
          <Button
            type="primary"
            size="small"
            data-testid="emit-click-2"
            onClick={() => signatureRef.current?.confirm()}
          >
            确认
          </Button>
        </div>
      </>
    )
  }
  const { container, getByTestId } = render(<App />)

  expect(container.querySelector('.nut-signature')).toHaveClass(
    'test-signature'
  )
  expect(container?.innerHTML).toMatchSnapshot()
  fireEvent.click(getByTestId('emit-click-2'))
  fireEvent.click(getByTestId('emit-click-1'))
})
