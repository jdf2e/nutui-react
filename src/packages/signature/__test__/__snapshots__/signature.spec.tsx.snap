// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props test 1`] = `"<div class=\\"nut-signature test-signature\\"><div class=\\"nut-signature__inner\\"><canvas height=\\"0\\" width=\\"0\\"></canvas></div></div>"`;
